### PR TITLE
91583 json to xml mapping merging ros events

### PR DESCRIPTION
**Pull request by sakshi15525**
1. Created new webjob project.
2. Added webjob configurations.
3. Code refactoring done.
4. Added helper and model class to send message in Azure queue.
5. Added mapping and merging logic for ros events.
6. Added logging.

### CODE REVIEW

**Review by pravinvarade10801 — COMMENTED** [review]
two minor comments.

File: src/UKHO.ERPFacade.Common/Configuration/AzureStorageConfiguration.cs [hunk 1]
```diff
@@ -6,5 +6,7 @@ namespace UKHO.ERPFacade.Common.Configuration
     public class AzureStorageConfiguration
     {
         public string ConnectionString { get; set; } = string.Empty;
+
+        public string QueueName { get; set; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (sakshi15525) [reply src/UKHO.ERPFacade.Common/Configuration/AzureStorageConfiguration.cs]: yes, removed

File: src/UKHO.ERPFacade.EventAggregation.WebJob/appsettings.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        "SapServiceOperationForRecordOfSale": "Z_ADDS_ROS",
+        "SapUsernameForRecordOfSale": "",
+        "SapPasswordForRecordOfSale": ""
+    },
+    "AzureWebJobsStorage": {
+        "ConnectionString": ""
+    },
+    "AzureStorageConfiguration": {
+        "ConnectionString": "",
+        "QueueName": ""
```
[the remark this answers is not in the document]
Comment (sakshi15525) [reply src/UKHO.ERPFacade.EventAggregation.WebJob/appsettings.json]: removed

File: src/UKHO.ERPFacade.Common/HttpClients/SapClient.cs [hunk 3]
```diff
@@ -18,24 +18,15 @@ public SapClient(HttpClient httpClient, IOptions<SapConfiguration> sapConfig)
             _sapConfig = sapConfig ?? throw new ArgumentNullException(nameof(sapConfig));
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.Common/HttpClients/SapClient.cs:18]: Please remove unused variable _sapConfig and related namespaces

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Services/AggregationService.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                    }
+                }
+                else
+                {
+                    _logger.LogWarning(EventIds.RequestAlreadyCompleted.ToEventId(), "The record has been completed already. | _X-Correlation-ID : {_X-Correlation-ID}", message.CorrelationId);
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(EventIds.UnhandledWebJobException.ToEventId(), ex, "Exception occured while processing Event Aggregation WebJob. | _X-Correlation-ID : {_X-Correlation-ID}", message.CorrelationId);
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Services/AggregationService.cs:98]: Spell mistake - occured, it should be occurred

**Review by pravinvarade10801 — COMMENTED** [review]
few minor comments

File: src/UKHO.ERPFacade.API/Controllers/WebhookController.cs [hunk 5]
```diff
@@ -23,38 +23,39 @@ public class WebhookController : BaseController<WebhookController>
...
         private const string LicenceUpdatedEventFileName = "LicenceUpdatedEvent.json";
         private const string RecordOfSaleContainerName = "recordofsaleblobs";
-        private const string RecordOfSaleEventFileName = "RecordOfSaleEvent.json";
+        private const string JsonFileType = ".json";
 
         public WebhookController(IHttpContextAccessor contextAccessor,
                                  ILogger<WebhookController> logger,
                                  IAzureTableReaderWriter azureTableReaderWriter,
                                  IAzureBlobEventWriter azureBlobEventWriter,
+                                 IAzureQueueMessaging azureQueueMessaging,
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.API/Controllers/WebhookController.cs:45]:
```suggestion
                                 IAzureQueueHelper azureQueueHelper,
```

File: src/UKHO.ERPFacade.API/Controllers/WebhookController.cs [hunk 6]
```diff
@@ -153,26 +155,12 @@ public virtual async Task<IActionResult> RecordOfSalePublishedEventReceived([Fro
...
-            {
-                _logger.LogError(EventIds.ErrorOccurredInSapForRecordOfSalePublishedEvent.ToEventId(), "An error occurred while sending record of sale event data to SAP. | {StatusCode}", response.StatusCode);
-                throw new ERPFacadeException(EventIds.ErrorOccurredInSapForRecordOfSalePublishedEvent.ToEventId());
-            }
-
-            _logger.LogInformation(EventIds.RecordOfSalePublishedEventDataPushedToSap.ToEventId(), "The record of sale event data has been sent to SAP successfully. | {StatusCode}", response.StatusCode);
-
-            await _azureTableReaderWriter.UpdateRecordOfSaleEventStatus(correlationId);
+            _logger.LogInformation(EventIds.AddMessageToAzureQueue.ToEventId(), "Adding the received Record of sale published event in queue storage.");
+            await _azureQueueMessaging.SendMessageToQueue(recordOfSaleEventJson);
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.API/Controllers/WebhookController.cs:162]:
```suggestion
            await _azureQueueMessaging.AddMessage(recordOfSaleEventJson);
```

File: src/UKHO.ERPFacade.Common/IO/Azure/AzureQueueMessaging.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+using Azure.Storage.Queues;
+using UKHO.ERPFacade.Common.Models.QueueEntities;
+
+namespace UKHO.ERPFacade.Common.IO.Azure
+{
+    [ExcludeFromCodeCoverage]
+    public class AzureQueueMessaging : IAzureQueueMessaging
+    {
+        private readonly IOptions<AzureStorageConfiguration> _azureStorageConfig;
+        private const string RecordOfSaleQueueName = "recordofsaleevents";
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.Common/IO/Azure/AzureQueueMessaging.cs:16]: can we keep queue name in appsettings.json file please?

File: src/UKHO.ERPFacade.Common/Models/QueueEntities/QueueMessageEntity.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Newtonsoft.Json;
+
+namespace UKHO.ERPFacade.Common.Models.QueueEntities
+{
+    [ExcludeFromCodeCoverage]
+    public class QueueMessageEntity
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.Common/Models/QueueEntities/QueueMessageEntity.cs:7]:
```suggestion
    public class RecordOfSaleQueueMessageEntity
```

File: src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        private readonly ILogger<EventAggregationWebjob> _logger;
+        private readonly IAggregationService _aggregationService;
+
+        public EventAggregationWebjob(ILogger<EventAggregationWebjob> logger, IAggregationService aggregationService)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _aggregationService = aggregationService ?? throw new ArgumentNullException(nameof(aggregationService));
+        }
+
+        public async Task ProcessQueueMessage([QueueTrigger("recordofsaleevents")] QueueMessage message)
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs:22]: Queue name should be retrieved from configuration.

File: src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            _logger.LogInformation(EventIds.WebjobForEventAggregationStarted.ToEventId(), "Webjob started for merging record of sale events.");
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs:24]:
```suggestion
            _logger.LogInformation(EventIds.WebjobForEventAggregationStarted.ToEventId(), "Webjob has started to process and merge sliced events.");
```

File: src/UKHO.ERPFacade.Common/IO/Azure/AzureBlobEventWriter.cs [hunk 11]
```diff
@@ -67,6 +67,23 @@ public bool DeleteContainer(string blobContainerName)
             return blobContainerClient.DeleteIfExists();
         }
 
+        public List<string> GetBlobNamesInFolder(string blobContainerName, string corrId)
```
[the remark this answers is not in the document]
Comment (sakshi15525) [reply src/UKHO.ERPFacade.Common/IO/Azure/AzureBlobEventWriter.cs:70]: No, here we are expecting different output.

File: src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            await _aggregationService.MergeRecordOfSaleEvents(message);
+            _logger.LogInformation(EventIds.WebjobForEventAggregationCompleted.ToEventId(), "Webjob completed for merging record of sale events.");
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/EventAggregationWebjob.cs:26]:
```suggestion
            _logger.LogInformation(EventIds.WebjobForEventAggregationCompleted.ToEventId(), "Webjob is completed");
```

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                    rosNewLicencePayload.EndDate = eventData.Data.RecordsOfSale.HoldingsExpiryDate;
+                    rosNewLicencePayload.LicenceType = eventData.Data.RecordsOfSale.LicenceType;
+                    rosNewLicencePayload.LicenceDuration = eventData.Data.RecordsOfSale.LicenceDuration;
+                    rosNewLicencePayload.FleetName = string.Empty;
+                    rosNewLicencePayload.LicenceNumber = string.Empty;
+
+                    PROD prod = new();
+                    List<UnitOfSales> unitOfSaleList = new();
+
+                    foreach (var rosUnitOfSale in eventData.Data.RecordsOfSale.RosUnitOfSale)
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:101]: Can you try converting this to LINQ expression?

File: src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+                    prod.UnitOfSales = unitOfSaleList;
+                    rosNewLicencePayload.PROD = prod;
+                }
+
+                else
+                {
+                    List<UnitOfSales> existingUnitOfSaleList = rosNewLicencePayload.PROD.UnitOfSales;
+
+                    foreach (var rosUnitOfSale in eventData.Data.RecordsOfSale.RosUnitOfSale)
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/Helpers/RecordOfSaleSapMessageBuilder.cs:123]: Same as above

File: src/UKHO.ERPFacade.EventAggregation.WebJob/appsettings.json [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        "SapUsernameForRecordOfSale": "",
+        "SapPasswordForRecordOfSale": ""
+    },
+    "AzureWebJobsStorage": {
+        "ConnectionString": ""
+    },
+    "AzureStorageConfiguration": {
+        "ConnectionString": ""
+    },
+    "QueuesOptions": {
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.EventAggregation.WebJob/appsettings.json:42]: These values should not be hard coded. Those has to be environment specific.

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few review comments.

**Review by BugShredder — APPROVED** [review]

**Review by Sadha1501493 — APPROVED** [review]
LGTM. Thanks @Vishal14583

File: src/UKHO.ERPFacade.Common/IO/Azure/AzureQueueMessaging.cs (hunk 7 above)
Comment (sakshi15525) [reply src/UKHO.ERPFacade.Common/IO/Azure/AzureQueueMessaging.cs]: Not required for now

**Review by Vishal14583 — APPROVED** [review]
LGTM

**Review by pravinvarade10801 — APPROVED** [review]
LGTM

**Review by jrippington — APPROVED** [review]
Looks good

**Review by JiviteshT — APPROVED** [review]
Looks good

**Review by HugoBurgess — APPROVED** [review]
Looks good

**Pull request merged**